modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/ExecutionDistributionProviderImpl.java

```diff
@@ -44,61 +47,83 @@
+            DistributionHolder holder = new DistributionHolderImpl(List.of(initiatorNode), Map.of(), Map.of());
+
+            return completedFuture(holder);
+        } else {
+            return distribution0(tables, views, operationTime, mapOnBackups, initiatorNode);
+        }
     }
 
-    @Override
-    public CompletableFuture<ExecutionTarget> forSystemView(ExecutionTargetFactory factory, IgniteSystemView view) {
-        List<String> nodes = systemViewManager.owningNodes(view.name());
-
-        if (nullOrEmpty(nodes)) {
-            return failedFuture(
-                    new SqlException(Sql.MAPPING_ERR, format("The view with name '{}' could not be found on"
-                            + " any active nodes in the cluster", view.name()))
-            );
+    private CompletableFuture<DistributionHolder> distribution0(
+            Collection<IgniteTable> tables,
+            Collection<String> views,
+            HybridTimestamp operationTime,
+            boolean mapOnBackups,
+            String initiatorNode
+    ) {
+        Map<IgniteTable, CompletableFuture<List<TokenizedAssignments>>> mapResult = newHashMap(tables.size());
+        Map<IgniteTable, List<TokenizedAssignments>> mapResultResolved = newHashMap(tables.size());
+
+        for (IgniteTable tbl : tables) {
+            CompletableFuture<List<TokenizedAssignments>> assignments = collectAssignments(tbl, operationTime, mapOnBackups);
+
+            mapResult.put(tbl, assignments);
         }
 
-        return completedFuture(
-                view.distribution() == IgniteDistributions.single()
-                        ? factory.oneOf(nodes)
-                        : factory.allOf(nodes)
+        CompletableFuture<Void> all = CompletableFuture.allOf(mapResult.values().toArray(new CompletableFuture[0]));
+
+        CompletableFuture<Map<IgniteTable, List<TokenizedAssignments>>> fut = all.thenApply(v -> mapResult.entrySet().stream()
+                .map(e -> Map.entry(e.getKey(), e.getValue().join()))
+                .collect(Collectors.toMap(Entry::getKey, Entry::getValue))
         );
+
+        CompletableFuture<List<String>> participantNodes = fut.thenApply(
+                v -> v.values().stream().flatMap(List::stream).flatMap(i -> i.nodes().stream()).map(Assignment::consistentId)
+                        .distinct()
+                        .collect(Collectors.toList()));
+
+        return participantNodes.thenApply(nodes -> {
+            nodes.add(initiatorNode);
+
+            // this is a safe join, because we have waited for all futures to be completed
+            mapResult.forEach((k, v) -> mapResultResolved.put(k, v.join()));
+
+            Map<String, List<String>> nodesPerView = views.stream().distinct()
+                    .collect(Collectors.toMap(Function.identity(), systemViewManager::owningNodes));
+
+            List<String> viewNodes = nodesPerView.values().stream().flatMap(List::stream).collect(Collectors.toList());
```

Comment:
will be better to use `toUnmodifiableList`

Reply:
done